Refactor `cont.wast` to not use `let` & `func.bind`

dhil opened the pull request:
This patch refactors the test `cont.wast` such that it no longer uses the obsolete `let` instruction.

dhil committed:
Refactor `cont.wast` to not use `let`

This patch refactors the test `cont.wast` such that it no longer uses
the obsolete `let` instruction.

dhil commented:
@rossberg @slindley PTAL. This patch brings us closer to be able to integrate `cont.wast` into wasm-tools/wasmtime testsuite. ~~We still need to rewrite the parts that use `func.bind`~~.

rossberg commented on test/core/cont.wast:
```diff
@@ -86,18 +86,18 @@
     (unreachable)
   )
   (func $nl3 (param $k (ref $k1))
+    (local $k' (ref null $k1))
```
This should work as well.
```suggestion
    (local $k' (ref $k1))
```

dhil replied:
I get the following error then
```
../test/core/cont.wast:89.16-89.25: invalid module: non-defaultable local type
```

rossberg commented on test/core/cont.wast:
```diff
@@ -214,18 +214,20 @@
   (func (export "sum") (param $i i64) (param $j i64) (result i64)
     (local $sum i64)
+    (local $n i64)
+    (local $k (ref null $cont))
```
```suggestion
    (local $k (ref $cont))
```

rossberg commented:
Oh, I suppose this repo hasn't been updated to the final locals semantics yet.

Would you be willing to merge from upstream function-references/main? I suspect there might be some conflicts, however...

dhil committed:
Refactor `cont.wast` to drop uses of `func.bind`.

This commit refactors the fiel `cont.wast` such that it no longer uses
`func.bind`. We change the type of `spawn` such that it is
parameterised by a continuation type rather than a function type, and
then we replace every use of `func.bind` by a corresponding
`cont.bind`.

dhil commented:
> Oh, I suppose this repo hasn't been updated to the final locals semantics yet.
> 
> Would you be willing to merge from upstream function-references/main? I suspect there might be some conflicts, however...

Yes. I can try to sync in changes from function-references and exception-handling. But I would like to do this in a separate PR.

My latest commit further refactors `cont.wast` such that it no longer uses `func.bind`. With these changes I believe I can integrate it into the wasmtime testsuite.

slindley reviewed (approved):
Looks fine to me
